conda_build/jinja_context.py

```diff
@@ -243,18 +243,23 @@ def pin_compatible(m, package_name, lower_bound=None, upper_bound=None, min_pin=
     return "  ".join((package_name, compatibility)) if compatibility is not None else package_name
 
 
-def pin_subpackage_against_outputs(key, outputs, min_pin, max_pin, exact, permit_undefined_jinja):
-    # If we can finalize the metadata at the same time as we create metadata.other_outputs then
-    # this function is not necessary and can be folded back into pin_subpackage.
-    pin = None
-    subpackage_name, _ = key
+def pin_subpackage_against_outputs(metadata, matching_package_keys, outputs, min_pin, max_pin,
+                                   exact, permit_undefined_jinja):
     # two ways to match:
     #    1. only one other output named the same as the subpackage_name from the key
     #    2. whole key matches (both subpackage name and variant)
-    keys = list(outputs.keys())
-    matching_package_keys = [k for k in keys if k[0] == subpackage_name]
+
     if len(matching_package_keys) == 1:
         key = matching_package_keys[0]
+    elif len(matching_package_keys) > 1:
+        key = None
+        for pkg_name, variant in matching_package_keys:
+            shared_vars = set(variant.keys()) & set(metadata.config.variant.keys())
```

Comment:
This matches other outputs with any keys that are common to both metadata objects.  For debugging, the keys are always the (package name, used vars+values).  It used to be (package name, variant) - but that was really big and hard to look at.

Reply:
The other important thing here that this PR fixes is that matching ALL of a variant was breaking things like

libxyz (vs2015)
py-xyz (py35, py36)

Only one of those py packages would match up, because the complete variant would only match once.  Now their common subset will be only the c_compiler, and it'll work.